File: src/main/java/graphql/schema/GraphQLFieldDefinition.java
package graphql.schema;


import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static graphql.Assert.assertNotNull;

public class GraphQLFieldDefinition<T>
{

    private final String name;
    private final String description;
    private GraphQLOutputType type;
    private final DataFetcher<T> dataFetcher;
    private final String deprecationReason;
    private final List<GraphQLArgument> arguments = new ArrayList<>();


    public GraphQLFieldDefinition(String name,
                                  String description,
                                  GraphQLOutputType type,
                                  DataFetcher<T> dataFetcher,
                                  List<GraphQLArgument> arguments,
                                  String deprecationReason)
    {
        assertNotNull(name, "name can't be null");
        assertNotNull(dataFetcher, "dataFetcher can't be null");
        assertNotNull(type, "type can't be null");
        assertNotNull(arguments, "arguments can't be null");
        this.name = name;
        this.description = description;
        this.type = type;
        this.dataFetcher = dataFetcher;
        this.arguments.addAll(arguments);
        this.deprecationReason = deprecationReason;
    }


    void replaceTypeReferences(Map<String, GraphQLType> typeMap)
    {
        type = (GraphQLOutputType)new SchemaUtil().resolveTypeReference(type, typeMap);
    }

    public String getName()
    {
        return name;
    }


    public GraphQLOutputType getType()
    {
        return type;
    }

    public DataFetcher<T> getDataFetcher()
    {
        return dataFetcher;
    }

    public GraphQLArgument getArgument(String name)
    {
        for (GraphQLArgument argument : arguments) {
            if (argument.getName().equals(name)) return argument;
        }
        return null;
    }

    public List<GraphQLArgument> getArguments()
    {
        return new ArrayList<GraphQLArgument>(arguments);
    }

    public String getDescription()
    {
        return description;
    }

    public String getDeprecationReason()
    {
        return deprecationReason;
    }

    public boolean isDeprecated()
    {
        return deprecationReason != null;
    }

    public static <T> Builder<T> newFieldDefinition()
    {
        return new Builder<>();
    }

    public static class Builder<T>
    {

        private String name;
        private String description;
        private GraphQLOutputType type;
        private DataFetcher<T> dataFetcher;
        private List<GraphQLArgument> arguments = new ArrayList<>();
        private String deprecationReason;
        private boolean isField;


        public Builder<T> name(String name)
        {
            this.name = name;
            return this;
        }

        public Builder<T> description(String description)
        {
            this.description = description;
            return this;
        }

        public Builder<T> type(GraphQLObjectType.Builder builder)
        {
            return type(builder.build());
        }

        public Builder<T> type(GraphQLInterfaceType.Builder builder)
        {
            return type(builder.build());
        }

        public Builder<T> type(GraphQLUnionType.Builder builder)
        {
            return type(builder.build());
        }

        public Builder<T> type(GraphQLOutputType type)
        {
            this.type = type;
            return this;
        }

        public Builder<T> dataFetcher(DataFetcher<T> dataFetcher)
        {
            this.dataFetcher = dataFetcher;
            return this;
        }

        public Builder<T> staticValue(final T value)
        {
            this.dataFetcher = environment -> CompletableFuture.completedFuture(value);
            return this;
        }

        /**
         * Get the data from a field, rather than a property.
         *
         * @return this builder
         */
        public Builder<T> fetchField()
        {
            this.isField = true;
            return this;
        }

        public Builder<T> argument(GraphQLArgument argument)
        {
            this.arguments.add(argument);
            return this;
        }

        /**
         * Take an argument builder in a function definition and apply. Can be used in a jdk8 lambda
         * e.g.:
         * <pre>
         *     {@code
         *      argument(a -> a.name("argumentName"))
         *     }
         * </pre>
         *
         * @param builderFunction a supplier for the builder impl
         * @return this
         */
        public Builder<T> argument(BuilderFunction<GraphQLArgument.Builder> builderFunction)
        {
            GraphQLArgument.Builder builder = GraphQLArgument.newArgument();
            builder = builderFunction.apply(builder);
            return argument(builder);
        }

        /**
         * Same effect as the argument(GraphQLArgument). Builder.build() is called
         * from within
         *
         * @param builder an un-built/incomplete GraphQLArgument
         * @return this
         */
        public Builder<T> argument(GraphQLArgument.Builder builder)
        {
            this.arguments.add(builder.build());
            return this;
        }

        public Builder<T> argument(List<GraphQLArgument> arguments)
        {
            this.arguments.addAll(arguments);
            return this;
        }

        public Builder<T> deprecate(String deprecationReason)
        {
            this.deprecationReason = deprecationReason;
            return this;
        }

        public GraphQLFieldDefinition<T> build()
        {
            if (dataFetcher == null) {
                if (isField) {
                    dataFetcher = new FieldDataFetcher<>(name);
                } else {
                    dataFetcher = new PropertyDataFetcher<>(name);
                }
            }
            return new GraphQLFieldDefinition<>(name, description, type, dataFetcher, arguments, deprecationReason);
        }


    }
}

File: src/main/kotlin/graphql/schema/GraphQLFieldDefinition.kt
package graphql.schema


import graphql.Assert.assertNotNull
import java.util.*
import kotlin.properties.Delegates

class GraphQLFieldDefinition<T>(val name: String,
                                val description: String?,
                                type: GraphQLOutputType,
                                val dataFetcher: DataFetcher<T>?,
                                val arguments: List<GraphQLArgument>,
                                val deprecationReason: String?) {
    var type: GraphQLOutputType by Delegates.notNull<GraphQLOutputType>()
        private set


    init {
        assertNotNull(type, "type can't be null")
        this.type = type
    }


    internal fun replaceTypeReferences(typeMap: Map<String, GraphQLType>) {
        type = SchemaUtil().resolveTypeReference(type, typeMap) as GraphQLOutputType
    }

    fun getArgument(name: String): GraphQLArgument? {
        for (argument in arguments) {
            if (argument.name == name) return argument
        }
        return null
    }

    val deprecated: Boolean
        get() = deprecationReason != null


    @GraphQLDslMarker
    class Builder<T> {
        var name: String by Delegates.notNull<String>()
        var description: String? = null
        var type: GraphQLOutputType by Delegates.notNull<GraphQLOutputType>()
        var fetcher: DataFetcher<T>? = null
        val arguments = ArrayList<GraphQLArgument>()
        var deprecationReason: String? = null
        var isField: Boolean = false


        fun name(name: String): Builder<T> {
            this.name = name
            return this
        }

        fun description(description: String): Builder<T> {
            this.description = description
            return this
        }

        fun type(builder: GraphQLObjectType.Builder): Builder<T> {
            return type(builder.build())
        }

        fun type(builder: GraphQLInterfaceType.Builder): Builder<T> {
            return type(builder.build())
        }

        fun type(builder: GraphQLUnionType.Builder): Builder<T> {
            return type(builder.build())
        }

        fun type(type: GraphQLOutputType): Builder<T> {
            this.type = type
            return this
        }

        fun fetcher(dataFetcher: DataFetcher<T>): Builder<T> {
            this.fetcher = dataFetcher
            return this
        }

        var staticValue : T? = null
            set (value){
                fetcher = staticDataFetcher(value)
            }

        fun staticValue(value: T): Builder<T> {
            this.fetcher = staticDataFetcher(value)

            return this
        }

        fun argument(argument: GraphQLArgument): Builder<T> {
            this.arguments.add(argument)
            return this
        }


        fun argument(block: GraphQLArgument.Builder.() -> Unit): Unit {
            arguments += newArgument(block)
        }


        fun argument(builder: GraphQLArgument.Builder): Builder<T> {
            this.arguments.add(builder.build())
            return this
        }

        fun argument(arguments: List<GraphQLArgument>): Builder<T> {
            this.arguments.addAll(arguments)
            return this
        }

        fun build(): GraphQLFieldDefinition<T> {
            val fetcher: DataFetcher<T> = fetcher ?:
                    if (isField) {
                        fieldDataFetcher<T>(name)
                    } else {
                        propertyDataFetcher<T>(name)
                    }

            return GraphQLFieldDefinition(name, description, type, fetcher, arguments, deprecationReason)
        }
    }

    companion object {
        @JvmStatic
        fun <T> newFieldDefinition(): Builder<T> {
            return Builder()
        }
    }
}

inline fun <reified T : Any> newField(block: GraphQLFieldDefinition.Builder<T>.() -> Unit): GraphQLFieldDefinition<T> {
    val builder = GraphQLFieldDefinition.newFieldDefinition<T>()
    typeResolve(T::class)?.let { builder.type = it }
    builder.block()
    return builder.build()
}
